editor/src/tool/tools/resize.rs

```diff
@@ -1,34 +1,54 @@
 use crate::input::keyboard::Key;
 use crate::input::{mouse::ViewportPosition, InputPreprocessor};
 use crate::message_prelude::*;
-use glam::{DAffine2, Vec2Swizzles};
+use crate::tool::snapping::SnapHandler;
+use crate::tool::DocumentMessageHandler;
+use glam::{DAffine2, DVec2, Vec2Swizzles};
 use graphene::Operation;
 
 #[derive(Clone, Debug, Default)]
 pub struct Resize {
 	pub drag_start: ViewportPosition,
 	pub path: Option<Vec<LayerId>>,
+	snap_handler: SnapHandler,
 }
 impl Resize {
-	pub fn calculate_transform(&self, center: Key, lock_ratio: Key, ipp: &InputPreprocessor) -> Option<Message> {
-		let mut start = self.drag_start;
-		let stop = ipp.mouse.position;
+	/// Starts a resize, assigning the snap targets and snapping the starting position.
+	pub fn start(&mut self, document: &DocumentMessageHandler, mouse_position: DVec2) {
+		let layers = document.all_layers_sorted();
+		self.snap_handler.start_snap(document, layers, &[]);
+		self.drag_start = self.snap_handler.snap_position(document, mouse_position);
+	}
 
-		let mut size = stop - start;
-		if ipp.keyboard.get(lock_ratio as usize) {
-			size = size.abs().max(size.abs().yx()) * size.signum();
-		}
-		if ipp.keyboard.get(center as usize) {
-			start -= size;
-			size *= 2.;
-		}
+	pub fn calculate_transform(&self, document: &DocumentMessageHandler, center: Key, lock_ratio: Key, ipp: &InputPreprocessor) -> Option<Message> {
+		if let Some(path) = &self.path {
+			let mut start = self.drag_start;
 
-		self.path.clone().map(|path| {
-			Operation::SetLayerTransformInViewport {
-				path,
-				transform: DAffine2::from_scale_angle_translation(size, 0., start).to_cols_array(),
+			let stop = self.snap_handler.snap_position(document, ipp.mouse.position);
+
+			let mut size = stop - start;
+			if ipp.keyboard.get(lock_ratio as usize) {
+				size = size.abs().max(size.abs().yx()) * size.signum();
+			}
+			if ipp.keyboard.get(center as usize) {
```

Comment:
Not at all your pr, just something I noticed:
Should we add a `ipp.keyboard.is_presseed(key: Key)` method?